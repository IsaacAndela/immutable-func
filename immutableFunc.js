/*eslint-disable max-len*/
import {Map, Set, List, Stack, Record, Seq, OrderedMap, OrderedSet} from 'immutable';
import {funcItUp} from './funcItUp.js';
/*eslint-enable max-len*/

const isFunction = (val) => typeof val === 'function';

const startsWithLetter = (str) => /^[a-zA-Z]/.test(str);

const constructors = List([
	Map,
	Set,
	List,
	Stack,
	Record,
	OrderedMap,
	OrderedSet,
	Seq,
	Seq.Indexed,
	Seq.Keyed,
	Seq.Set,
]);

const prototypes = constructors.map((constructor) => constructor.prototype);

// console.log('prototypes', prototypes);

const filterFunctions = (obj) => {
	let memo = Set();

	for (let key in obj) {
		const val = obj[key];

		if (isFunction(val) && startsWithLetter(key)) {
			memo = memo.add(key);
		}
	}

	return memo;
};

export const functionKeys = prototypes.reduce((memo, prototype) => (
	memo.union(filterFunctions(prototype))
), Set()).sort();

const immutableFunc = functionKeys.reduce((obj,
	functionKey) => {
	obj[functionKey] = funcItUp(functionKey);

	return obj;
}, {});

export default immutableFunc;
