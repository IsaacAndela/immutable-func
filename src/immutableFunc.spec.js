import * as funcs from '../immutableFunc';

it('should have converted all methods to functions', () => {
	expect(Object.keys(funcs)).toMatchSnapshot();
});
