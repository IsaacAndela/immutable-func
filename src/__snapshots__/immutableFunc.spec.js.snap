exports[`test should have converted all methods to functions 1`] = `
Array [
  "add",
  "asImmutable",
  "asMutable",
  "butLast",
  "cacheResult",
  "chain",
  "clear",
  "concat",
  "contains",
  "count",
  "countBy",
  "deleteIn",
  "entries",
  "entrySeq",
  "equals",
  "every",
  "filter",
  "filterNot",
  "find",
  "findEntry",
  "findIndex",
  "findKey",
  "findLast",
  "findLastEntry",
  "findLastIndex",
  "findLastKey",
  "first",
  "flatMap",
  "flatten",
  "flip",
  "forEach",
  "fromEntrySeq",
  "get",
  "getIn",
  "groupBy",
  "has",
  "hasIn",
  "hashCode",
  "includes",
  "indexOf",
  "insert",
  "inspect",
  "interleave",
  "interpose",
  "intersect",
  "isEmpty",
  "isSubset",
  "isSuperset",
  "join",
  "keyOf",
  "keySeq",
  "keys",
  "last",
  "lastIndexOf",
  "lastKeyOf",
  "map",
  "mapEntries",
  "mapKeys",
  "max",
  "maxBy",
  "merge",
  "mergeDeep",
  "mergeDeepIn",
  "mergeDeepWith",
  "mergeIn",
  "mergeWith",
  "min",
  "minBy",
  "peek",
  "pop",
  "push",
  "pushAll",
  "reduce",
  "reduceRight",
  "remove",
  "removeIn",
  "rest",
  "reverse",
  "set",
  "setIn",
  "setSize",
  "shift",
  "skip",
  "skipLast",
  "skipUntil",
  "skipWhile",
  "slice",
  "some",
  "sort",
  "sortBy",
  "splice",
  "subtract",
  "take",
  "takeLast",
  "takeUntil",
  "takeWhile",
  "toArray",
  "toIndexedSeq",
  "toJS",
  "toJSON",
  "toKeyedSeq",
  "toList",
  "toMap",
  "toObject",
  "toOrderedMap",
  "toOrderedSet",
  "toSeq",
  "toSet",
  "toSetSeq",
  "toSource",
  "toStack",
  "toString",
  "union",
  "unshift",
  "unshiftAll",
  "update",
  "updateIn",
  "valueSeq",
  "values",
  "wasAltered",
  "withMutations",
  "zip",
  "zipWith",
]
`;
